core/doc/man/Vaile.1
.TH VAILE 1 2020-03-23 ""

.SH NAME
.P
\fBVaile\fR — Metasploit\-like pentest framework derived from TIDoS

.SH SYNOPSIS
.P
\fBVaile\fR [\fIOPTION\fR]...

.SH DESCRIPTION
.P
\fBVaile\fR is a versatile collection of tools for \fIwebapp\fR security testing. It was designed in order to help penetration testers during all phases of an operation, especially reconnaissance, scanning and vulnerability analysis.

.P
It features 3 easy\-to\-use interfaces: an interactive shell, an inline CLI and a graphical user interface powered by Qt. Its modules are easily comprehensible and fully verbose, and are using \fImultiprocessing\fR to make attacking more efficient. If requested, Vaile will also pipe attacks through \fITor\fR, rendering your attacks more anonymous and difficult to track.

.P
The \fBVaile\fR framework is module\-based, so users can easily create their own modules, or edit the built\-in ones, respecting the general structure of the already existent modules.

.P
Its structure allows it to evolve and be easily adapted to new features and modules. Currently, it has 108 modules ready to be used.

.SH OPTIONS
.P
Running \fBVaile\fR without any argument runs an interactive, metasploit\-like shell, with all settings set to default.

.TP
You can also specify a target URL, combined with a module, to execute this module on given target in a quicker and easier way.
\fB$ Vaile \-v http://example.com \-l pathtrav\fR

.TP
You also can use a previously saved \fIsession\fR file as argument, to use it instead of a target URL. Sessions can (and need to) be saved with the \fBsessions\fR command.
\fB$ Vaile \-v saarsec \-l pathtrav \-s\fR

.TP
Using Vaile's \fBVAL\fR file format, you can fully automate attacks, since it sets targets, modules and options automatically. Load a file up this way: 
\fB$ Vaile \-c syn.val\fR

.TP
When attacking per CLI, you can also make use of the \fBtor\fR command to pipe attacks through the Tor network, like this:
\fB$ Vaile \-v http://example.com \-l pathtrav \-\-tor\fR

.TP
Finally, \fBVaile\fR can be run as a graphical user interface, using and wrapping around the core framework.
\fB$ Vaile \-\-app\fR

.TP
For inline help and more options, use \fBVaile \-\-help\fR

.SH INTERFACE
.P
The main interface of \fBVaile\fR framework is its interactive, metasploit\-like console.

.P
\fBTO GET HELP\fR on a Vaile command, you can use "\fBhelp <command>\fR" or "\fB? <command>\fR". To list all available commands, use "\fBhelp\fR" resp. "\fB?\fR" without argument. This will also give a short summary on the commands, as listed below.

.P
\fICommands:\fR

.TP
\fBattack\fR
Attack specified target(s)
.TP
\fBclear\fR
Clear terminal
.TP
\fBcreds\fR
Handle target credentials
.TP
\fBfetch\fR
Check for and install updates
.TP
\fBfind\fR
Search a module
.TP
\fBhelp\fR
Show help message
.TP
\fBinfo\fR
Show description of current module
.TP
\fBintro\fR
Display Intro
.TP
\fBleave\fR
Leave module
.TP
\fBlist\fR
List all modules of a category
.TP
\fBload\fR
Load module
.TP
\fBnetinfo\fR
Show network information
.TP
\fBopts\fR
Show options of current module
.TP
\fBphpsploit\fR
Load the phpsploit framework (needs to be downloaded externally)
.TP
\fBprocesses\fR
Set number of processes in parallelisation
.TP
\fBq\fR
Terminate Vaile session
.TP
\fBsessions\fR
Interact with cached sessions
.TP
\fBset\fR
Set option value of module
.TP
\fBtor\fR
Pipe Attacks through the Tor Network
.TP
\fBvicadd\fR
Add Target to list
.TP
\fBvicdel\fR
Delete Target from list
.TP
\fBviclist\fR
List all targets

.SH ISSUES
.TP
To submit any issue, bug or proposal, please send it in Vaile's issues section:
https://github.com/VainlyStrain/Vaile/issues

.SH CONTRIBUTE
.TP
If you want to contribute to \fBVaile\fR, submit a module, patch, or anything else, take a look at the \fBCONTRIBUTING\fR file, in the core/doc directory

.SH LICENCE
.P
This software is under the GNU GENERAL PUBLIC LICENSE Version 3, 29 June 2007


.\" man code generated by txt2tags 2.5 (http://txt2tags.sf.net)
.\" cmdline: txt2tags -q -t man -i man.txt2tags -o Vaile.1



files/CHANGELOG
2.0.0 (INIT)
--------------

  Conversion to Python3
  New console interface

2.0.1-3
---------

  Created separate file for handling credentials
  added temporar credential dropping in case of socket.gaierror

INTERMEDIATE

  phpsploit checks if path to script is valid ; else it will ask for a valid path
  nmapmain: omitted unnecessary display

2.0.1-5
---------

  fixed: "can't cd into tmp/logs"
  fixed: XSS byte-like object is required
  fixed: PHPi xrange
  improved: automatic dropping of credentials for OSINT-passive modules where they are not needed
  fixed: autocorrect strikes again in legal disclaimer ;)
 
2.0.1-15
----------

  fixed: CRLF
  fixed: Exploit invalid module error
  fixed: xrange in some modules
  fixed: string.ascii_letters in some modules
  fixed: Broken exception syntax in some modules
  fixed: SQLi error when leaving module
  improved: show content of found paths in pathtrav module
  improved: auto modules will carry on with next module when current one fails
  improved: banners (main, modules)
  fixed: XPATHi
  fixed: LFI
  fixed: HTMLi
  partially fixed: RCE
  improved: added support for multiple parameters in various modules
  improved: pathtrav evasion implemented in LFI, too
  
INTERMEDIATE

  fixed: osdetect bytes -> string error
  improved: vicadd & IPs
  new module: arpscan


[·= VERSION 2.1 - parallelisation ; tor ; options =·]

2.1.0-5
---------

  added: parallelisation mode for multiple modules to speed up calculation
  #all port scanners (parallelisation-only)
  #most of vlnysis-severe (choice through user input)
  added: "processes" command to set processes to be used in parallelisation
  added: new core methods to handle parallelisation
  fixed: TCP stealth scanner unable to find ports
  fixed: rce crash
  fixed: all subdom modules
  fixed: module x has no y in csrf
  fixed: payload duplication in sqli and xss
  fixed: getcensys
  improved: list&find only dispay name, not path
  improved: interface in "ip" controlled by variable in variables.py
  improved: non-verbose port scanning is less verbose
  
2.1.0-15
----------

  added: CLI interface for faster interaction with one specific module
         !!!you need to update your /bin/Vaile!!!
         new version in tmp/Vaile
  added: tor command to pipe attacks through the Tor network
         right now, only for vlnysis-severe (except CSRF, subdom)
  fixed: pathtrav exception during not-evasion attack

2.1.0-51
----------

  added: set-options for all vlnysis-severe modules and port scanners
  added: new .val file format for sessions with module properties
  added: "fetch" command to check and install updates
  added: cli option for fetch
  added: parallelisation for more modules
  added: Tor support for Web crawlers and other scanning modules (port scanners excluded)
  added: Tor and options for OSINT modules
  added: stub for upcoming shellcraft module
  added: parallelisation and options for finscan
  fixed: payload duplication in XSS
  fixed: placed finscan in portscanning category
  fixed: an error in listsplit potentially causing parallelised attacks to crash
  fixed: Brute Recon method already defined
  improved: target not necessary when using most aid modules
  improved: list command divides output in different categories, and sorts it alphabetically
  
2.1.5-0
---------

  fixed: import error in automation modules
  added: Tor for almost all modules

2.2.5-1
---------

  added: GUI
  added: full automation support (see the --file argument)
  several fixes
  improved: coloring, asciiart


Vaile.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
_____, ___
   '+ .;    
    , ;   
     .   
           
       .    
     .;.    
     .;  
      :  
      ,   
       

┌─[Vaile]─[]
└──╼ VainlyStrain
"""

import importlib as imp
import os
import platform
import sys
import time
from cmd import Cmd
import argparse

import core.methods.inputin as addtarget
import core.methods.print as prnt
import core.methods.select as select
import core.variables as varis
from core.Core.colors import R, B, C, color
from core.methods.cache import load, save, sessionparse
from core.methods.creds import creds
from core.methods.tor import torpipe, initcheck, session
from core.methods.parser import build_parser


class VainShell(Cmd):
    # prompt = "\033[1;31m──·»\033[0m\033[4mVaile\033[0m\033[1;31m]─[\033[0m{}\033[1;31m]\033[0m\033[1m –› \033[
    # 0m".format(varis.module)
    intro = ""
    prompt = '\033[0m\033[1m vaile > '
    ruler = "—-"
    doc_header = "Docvmented:"
    misc_header = "Misc.:"
    undoc_header = "NoDocs:"

    def cmdloop(self, intro=None):
        print(self.intro)
        while True:
            try:
                super(VainShell, self).cmdloop(intro="")
                break
            except KeyboardInterrupt:
                if varis.module == "":
                    print("^C\n" + R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Command 'q' to end session.")
                else:
                    print("^C")
                    
    def do_help(self, arg):
        """List available commands with "help" or detailed help with "help cmd"."""
        if arg:
            # XXX check arg syntax
            try:
                func = getattr(self, 'help_' + arg)
            except AttributeError:
                try:
                    doc = getattr(self, 'do_' + arg).__doc__
                    if doc:
                        self.stdout.write("%s\n" % str(doc))
                        return
                except AttributeError:
                    pass
                self.stdout.write("%s\n" % str(self.nohelp % (arg,)))
                return
            func()
        else:
            prnt.info()

    def help_help(self):
        print("""
  help [?]
  ----------

  Vaile help menu.
  Syntax: ? [CMD]

    CMD: command for which more detailed help should be shown.

  If CMD is omitted, a list of all commands with its function will be shown.
""")

    def do_q(self, inp) -> bool:
        return True

    def help_q(self):
        print("""
  q
  ---

  Terminate current session and quit the program.
  [!] The session is not cached, use command 'sessions' for this.
""")

    def sessionhelper(self, inp):
        print()
        options = sessionparse(inp)[1]
        for module, props in options.items():
            self.do_load(module)
            print("{}{}{}\n".format(color.UNDERLINE, module, C))
            for opt, val in props.items():
                self.do_set("{} {}".format(opt, val))
            self.do_leave("")
            print()

    def do_sessions(self, inp):
        if "load" in inp:
            b = varis.targets
            om = varis.module
            varis.targets = []
            #for i in varis.targets:
            #    varis.targets.remove(i)
            session = inp.split("load")[1].strip()
            if session == "":
                print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Syntax: sessions [load|save <SESS_ID>] [list]")
                varis.targets = b
            else:
                try:
                    if ".val" in session or "--val" in session:
                        session = session.replace("--val","").strip()
                        #print(session)
                        self.sessionhelper(session)
                        print(" [+] Restored VAL session: {}".format(session))
                        if om is not "":
                            self.do_load(om)
                    else:
                        load(session)
                        print(" [+] Restored session: {}.".format(session))
                except FileNotFoundError:
                    print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "{}: no such session file.".format(session))
                    varis.targets = b
        elif "save" in inp:
            session = inp.split("save")[1].strip()
            if session == "":
                print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Syntax: sessions [load|save <SESS_ID>] [list]")
            else:
                save(session)
        elif "list" in inp:
            os.system("\\ls core/sessioncache")
        else:
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Syntax: sessions [load|save <SESS_ID>] [list]")


    def help_sessions(self):
        print("""
  sessions
  ----------

  Interact with cached sessions.
  Available commands:

    list     list all available sessions to load.
    load ID  restore session ID
    save ID  save current session as ID.
""")

    def do_clear(self, inp):
        os.system('clear')

    def help_clear(self):
        print("""
  clear
  -------

  Clear the terminal using the native 'clear' command.
""")

    def do_tor(self, inp):
        try:
            initv = varis.initip == ""
            acc = False
            if initv:
                try:
                    initcheck()
                    acc = True
                except:
                    acc = False
            if "on" in inp.lower():
                if acc or not initv:
                    p = torpipe(True)
                    if p:
                        print(" [+] Tor > ON")
                    else:
                        varis.tor = False
                        start = input(" [?] Do you want to start the Tor service? (enter if not) :> ")
                        if start is not "":
                            try:
                                os.system("systemctl start tor")
                                print(" [+] Tor service successfully started.")
                                self.do_tor("on")
                            except Exception as e:
                                print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Starting Tor service failed:"+"\033[0m"+ color.CURSIVE +"\n{}".format(e) + C)
                else:
                    print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Starting Tor service failed: Initial IP not set."+"\033[0m" + C)
            elif "off" in inp.lower():
                torpipe(False)
                stop = input(" [?] Do you want to stop the Tor service? (enter if not) :> ")
                if stop is not "":
                    try:
                        os.system("systemctl stop tor")
                        print(" [+] Tor service successfully stopped.")
                    except Exception as e:
                        print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Stopping Tor service failed:"+"\033[0m"+ color.CURSIVE +"\n{}".format(e) + C)
                print(" [+] Tor > OFF")
            else:
                print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Syntax: tor on|off")
        except:
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Tor connection failed: IPcheck service not available.")

    def help_tor(self):
        print("""
  tor
  -----

  Pipe attacks over the Tor Anonymity Network.
  Syntax:

    tor on|off
""")

    def do_ip(self, inp):
        try:
            import core.methods.netinfo as netinfo
            netinfo.info()
        except:
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Something went wrong. Try again later.")

    def help_ip(self):
        print("""
  ip
  ----

  Provides current network information, such as

    + your local IP
    + your public IP
    + your MAC address
""")

    def do_creds(self, inp):
        creds(inp)
        
    def help_creds(self):
        print("""
  creds
  -------

  Add or remove credentials from a specific target.
  Syntax: creds add|del target

    target: the URL which shall be operated on

  [!] the target must be formatted as in viclist.
""")

    def do_find(self, inp):
        if inp == "":
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Please enter a search term.")
        else:
            select.search(inp)

    def help_find(self):
        print("""
  find
  ------

  Search a module by providing a search term.
  The command will return any module containing the term in

    + its name
    + its description
""")

    def do_list(self, inp):
        select.list(inp,True)

    def help_list(self):
        print("""
  list
  ------

  List all modules in a specified category.
  Providing no category, list will output all availbale categories.
  Available categories:

    all           all available modules
    aid           additional tools (e.g. honeypot check)
    infdisc       information disclosure modules
    osint-active  modules for active reconnaissance
    osint-passive modules for passive reconnaissance
    scan          modules for scanning and enumeration
    sploit        exploits (in progress)
    vlnysis       modules useful for vulnerability analysis
""")

    def do_attack(self, inp):
        if varis.module == "":
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "No module loaded.")
            return None
        elif "arpscan" in varis.module or "shellcraft" in varis.module or "encodeall" in varis.module or "hashes" in varis.module or "imgext" in varis.module:
            select.attack("")
        elif len(varis.targets) <= 0:
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "No target(s) set.")
            return None
        else:
            #print("\n Attack")
            #print(" --------")
            for i in varis.targets:
                if len(varis.targets) > 1:
                    print( "\n [i] Target: {}\n".format(i))
                select.attack(i)

    def help_attack(self):
        print("""
  attack
  --------

  unleash the loaded module on the specified target(s)
  if no options have been specified

    + default options will be applied
    + if above not possible, user will be prompted for live input.
""")

    def do_vicadd(self, inp):
        if "--ip" in inp:
            ip = inp.replace("--ip", "").strip()
            addtarget.inputip(ip)
        else:
            addtarget.inputin(inp)

    def help_vicadd(self):
        print("""
  vicadd
  --------

  Add the specified target to the target list.
  Syntax: viacdd [--ip] TARGET

    TARGET: the target added to the list.
    
  Options:
    --ip: specified target is an IP, rather than an URL
""")

    def do_phpsploit(self, inp):
        try:
            def filecheck():
                if not os.path.exists(varis.phpsploit):
                    print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "No phpsploit installation under {}".format(varis.phpsploit) + color.END)
                    phpsplt = input(" [#] Enter path to phpsploit script :> ")
                    varis.phpsploit = phpsplt
                    filecheck()
            filecheck()
            if inp == "":
                os.system("python3 {}".format(varis.phpsploit))
            else:
                os.system("sudo -u {} python3 {}".format(inp, varis.phpsploit))
        except SystemExit:
            pass
        except:
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "phpsploit crashed.")

    def help_phpsploit(self):
        print("""
  phpsploit
  -----------

  Load the phpsploit post-exploitation and control framework.
  Syntax: phpsploit [user]

    user  the user who will execute phpsploit (e.g. a non-privileged user)
  
  [!] you need to change the phpsploit path in core/variables.py to point at your installation.
""")

    def do_vicdel(self, inp):
        if inp == "all":
            varis.targets = []
            print(" [+] Cleared target list.")
        else:
            old = varis.targets
            varis.targets = list(filter(lambda a: a != inp, varis.targets))
            found = old != varis.targets
            if found:
                print(" [+] Deleted Target: {}".format(inp))
            else:
                print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Could not find specified target: {}".format(inp))

    def do_viclist(self, inp):
        for i in varis.targets:
            print(i)

    def help_viclist(self):
        print("""
  viclist
  ---------

  List all targets specified for attack.
""")

    def do_intro(self, inp):
        prnt.banner()
        prnt.bannerbelownew()
        print()

    def help_intro(self):
        print("""
  intro
  -------

  Display the intro banner.
""")

    def help_vicdel(self):
        print("""
  vicdel
  --------

  Remove a target from the list.
  Syntax: vicdel TARGET

    TARGET  target to be removed

  To delete all targets, use TARGET = all.
""")

    def emptyline(self):
        pass

    def do_set(self, inp):
        listed = inp.split(" ")
        if len(listed) != 2:
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Entry must contain exactly 1 space.")
        else:
            param = listed[0]
            value = listed[1]
            if varis.module != "":
                select.set(varis.module, param, value)
            else:
                print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "No module loaded.")

    def help_set(self):
        print("""
  set
  -----

  Set an attack option for the current module.
  Syntax: set OPT VAL

    OPT  name of the option to be modified
    VAL  value the option will take

  [!] VAL "none" for no value ("")
""")

    def do_info(self, inp):
        if varis.module == "":
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "No module loaded.")
        else:
            select.information(varis.module)

    def help_info(self):
        print("""
  info
  ------

  Displays the description of the current module, as well as all available options.
""")

    def do_opts(self, inp):
        if varis.module == "":
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "No module loaded.")
        else:
            select.opts(varis.module)

    def help_opts(self):
        print("""
  opts
  ------

  Displays the options of the current module.
""")

    def do_load(self, inp):
        try:
            success = False
            impmod = inp
            if "modules" in impmod:
                imp.import_module(impmod)
                success = True
            else:
                p = select.bareimport(impmod)
                success = p[0]
                impmod = p[1]
                imp.import_module(impmod)
            if success:
                varis.module = impmod
                self.prompt = '\033[0m\033[1m vaile({}{}\033[0m\033[1m) > '.format(R, varis.module.split(".")[-1])
        except ImportError:
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Not a valid module: {}".format(inp))
        except ValueError:
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Please enter a module.")

    def help_load(self):
        print("""
  load
  ------

  Load a module.
  Syntax: load MODULE

    MODULE: module to be loaded

  The full path, as well as the name can be used ('.' as separator)
""")

    def do_processes(self, inp):
        try:
            p = int(inp.strip())
            assert p > 0
            print(" [+] Processes: {} > {}".format(varis.processes, p))
            varis.processes = p
        except (ValueError, AssertionError):
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Not a valid (positive) integer: {}".format(inp))

    def help_processes(self):
        print("""
  processes
  -----------

  Sets the number of processes used in parallelised mode (default=5)
  Syntax: processes INT

    INT: a strictly positive integer
""")

    def do_leave(self, inp):
        varis.module = ""
        self.prompt = '\033[0m\033[1m vaile > '

    def help_leave(self):
        print("""
  leave
  -------

  Leave the current module.
""")

    do_EOF = do_q

    def do_fetch(self, inp):
        try:
            localver = varis.e_version.split("#")[0]
            s = session()
            onver = s.get("https://raw.githubusercontent.com/VainlyStrain/Vaile/master/core/doc/version").text.strip()
            localmain = localver.split("-")[0]
            localrev = localver.split("-")[1]
            locallist = localmain.split(".")
            onmain = onver.split("-")[0]
            onrev = onver.split("-")[1]
            onlist = onmain.split(".")
            uptodate = True
            for i in range(0, len(locallist)):
                if int(locallist[i]) < int(onlist[i]):
                    uptodate = False
            if uptodate:
                if int(localrev) < int(onrev):
                    uptodate = False
            if not uptodate:
                print(" [!] An update is available! Last version is: {}, installed version is: {}.".format(onver, localver))
                d = input(" [?] Do you want to update the framework? (enter if not) :> ")
                if d is not "":
                    path = os.path.dirname(os.path.realpath(__file__))
                    if "/home/" in path:
                        user = path.split("/")[2]
                        os.system("sudo -u {} git pull".format(user))
                    else:
                        os.system("git pull ; cp tmp/Vaile /bin/Vaile ; chmod +x /bin/Vaile")
                    print(" [+] Update installed successfully.")
            else:
                print(" [+] You are running the latest version of Vaile-framework ({}).".format(localver))
        except:
            print(R + " [-] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "An error occurred fetching...")

    def help_fetch(self):
        print("""
  fetch
  -------

  Check for and install updates of the framework.
""")

# help_EOF = help_q

def main():
    parser = build_parser()
    opt = vars(parser.parse_args())
    args = parser.parse_args()
    os.system('clear')
    if str(platform.system()) != "Linux":
        sys.exit(
            R + " [!] " + color.UNDERLINE + "\033[1m" + "You are not using a Linux Based OS! Linux is a must-have for "
                                                        "this script!" + color.END)
    if not os.geteuid() == 0:
        sys.exit(R + " [!] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Must be run as root." + B + " :)" + color.END)
    if 'no' in open('core/doc/choice').read():
        prnt.disclaimer()

        a1 = input(B + ' [?] Do you agree to these terms and conditions? :> ' + C)
        if a1.lower().startswith('y'):
            print(B + ' [+] That\'s awesome! Move on...')
            time.sleep(3)
            FILE = open("core/doc/choice", "w")
            FILE.write('yes')
            FILE.close()

        else:
            print(R + ' [!] ' + "\033[0m" + color.UNDERLINE + "\033[1m" + 'You have to agree!' + color.END)
            time.sleep(1)
            sys.exit(0)


    if opt["load"] and opt["victim"] and not opt["help"] and not opt["list"]:
        s = VainShell()
        if not opt["quiet"]:
            prnt.banner()
            prnt.bannerbelownew()
        if not opt["session"]:
            s.do_vicadd(args.victim)
        else:
            s.do_sessions("load {}".format(args.victim))
        s.do_load(args.load)
        if opt["tor"]:
            s.do_tor("on")
            if varis.tor:
               s.do_attack("")
        else:
            s.do_attack("")
    elif opt["help"]:
        if not opt["quiet"]:
            prnt.banner()
        #prnt.bannerbelownew()
        parser.print_help()
    elif opt["list"]:
        s = VainShell()
        if not opt["quiet"]:
            prnt.banner()
        s.do_list(args.list)
    elif opt["fetch"]:
        s = VainShell()
        s.do_fetch("")
    elif opt["victim"] and not opt["load"] or opt["load"] and not opt["victim"]:
        parser.error("'VIC' and 'M' are required for CLI attack.")
    else:
        if not opt["quiet"]:
            prnt.loadstyle()
            prnt.banner()
            prnt.bannerbelownew()
        VainShell().cmdloop()
        #print(R + "[Vaile] " + "\033[0m" + color.UNDERLINE + "\033[1m" + "Alvida, my friend!" + color.END)
        print(R + "[Vaile] " + "\033[0m" + color.END + "Alvida, my chosen")

if __name__ == '__main__':
    main()